src-java/nbworker-topology/nbworker-storm-topology/src/main/java/org/openkilda/wfm/topology/nbworker/validators/PathValidator.java

```diff
@@ -0,0 +1,513 @@
+        return getBandwidthWithReusableResources(inputData, isl,
+                pathSegment -> isSameReversePathSegment(inputData.getSegment(), pathSegment));
+    }
+
+    private long getBandwidthWithReusableResources(InputData inputData, Isl isl,
+                                                   Predicate<PathSegment> pathSegmentPredicate) {
+        if (inputData.getPath().getReuseFlowResources() != null
+                && !inputData.getPath().getReuseFlowResources().isEmpty()) {
+
+            Optional<Flow> flow = flowRepository.findById(inputData.getPath().getReuseFlowResources());
+
+            Optional<PathSegment> segment = flow.flatMap(value -> value.getPaths().stream()
+                    .map(FlowPath::getSegments)
+                    .flatMap(List::stream)
+                    .filter(pathSegmentPredicate)
+                    .findAny());
+
+            return segment.map(s -> isl.getAvailableBandwidth() + s.getBandwidth())
+                    .orElseGet(isl::getAvailableBandwidth);
+        }
+
+        return isl.getAvailableBandwidth();
+    }
+
+    private Set<String> validateEncapsulationType(InputData inputData, RepositoryData repositoryData) {
+        Set<String> errors = Sets.newHashSet();
+        Map<SwitchId, SwitchProperties> switchPropertiesMap = switchPropertiesRepository.findBySwitchIds(
+                Sets.newHashSet(inputData.getPath().getSrcSwitchId(), inputData.getPath().getDestSwitchId()));
+
+        if (!switchPropertiesMap.containsKey(inputData.getPath().getSrcSwitchId())) {
+            errors.add(getSrcSwitchNotFoundError(inputData));
+        } else {
+            if (!switchPropertiesMap.get(inputData.getPath().getSrcSwitchId()).getSupportedTransitEncapsulation()
+                    .contains(inputData.getPath().getFlowEncapsulationType())) {
+                errors.add(getSrcSwitchDoesNotSupportEncapsulationTypeError(inputData));
+            }
+        }
+
+        if (!switchPropertiesMap.containsKey(inputData.getPath().getDestSwitchId())) {
+            errors.add(getDestSwitchNotFoundError(inputData));
+        } else {
+            if (!switchPropertiesMap.get(inputData.getPath().getDestSwitchId()).getSupportedTransitEncapsulation()
+                    .contains(inputData.getPath().getFlowEncapsulationType())) {
+                errors.add(getDestSwitchDoesNotSupportEncapsulationTypeError(inputData));
+            }
+        }
+
+        return errors;
+    }
+
+    private Set<String> validateDiverseWithFlow(InputData inputData, RepositoryData repositoryData) {
+        if (!repositoryData.getIslRepository().findByEndpoints(inputData.getSegment().getSrcSwitchId(),
+                inputData.getSegment().getSrcPort(),
+                inputData.getSegment().getDestSwitchId(),
+                inputData.getSegment().getDestPort()).isPresent()) {
+            return Collections.singleton(getNoForwardIslError(inputData));
```

Comment:
Why are we checking ISL existence here? didn't we check it in validateForwardAndReverseLinks?

Reply:
I thought it's better to validate it than fail with some exception and don't need to make an assumption that some other validation has been executed before this one.

Reply:
okay